README
Tk::Clock
    Clock widget for Perl/Tk with analog and/or digital display

DESCRIPTION
    A canvas clock widget with both analog and digital represen-
    tation of both time and date.

    It is highly configurable for as far as colors and fonts and
    the way time and date is displayed.

    The year is displayed in two digits  in the default configu-
    ration, which makes it not Y2K compliant.

PREREQUISITES
    perl5 (probably 5.004.04 or better)
    Tk (probably 400.xxx or better)

    This port was tested on several OS's with perl-5.8.8-dor,
    perl-5.8.7-dor, and perl-5.8.5-dor in 32bit and 64bit builds
    all with Tk-804.027.

BUILDING AND INSTALLATION
    As all CPAN modules:
	# perl Makefile.PL
	# make test
	# make install

    make test should run the clock with some changing contrasting
    colors showing the following configurations for 4 seconds each:
    	1. Both analog and digital clock,
    	2. Only analog clock,
    	3. Only digital clock,
    	4. Both analog- and digital clock using an american like
    	   time/date format (for wich I shouldn't actualy support
    	   formats, cause amaricans do often not support european
    	   formats, but I thought, I'd be nice for this time),
	5. Only digital clock with an empty time format and a
	   double line date format showing the day of the week
	   and the date.

CHANGES
    See ChangeLog

COPYRIGHT AND LICENSE
    Copyright (C) 1999-2007 H.Merijn Brand

    This library is free software; you can redistribute it and/or modify
    it under the same terms as Perl itself. 

As always, have the appropriate amount of fun!


ChangeLog
0.37	- 02 Jan 2015, H.Merijn Brand
    * Use 12:15 AM instead of 00:15 AM
    * Upped Copyright to 2015

0.36	- 10 Apr 2014, H.Merijn Brand
    * Fix locales for non-utf8 (and windows)
    * Upped Copyright to 2014
    * Make sure boolean attributes are 0 or 1
    * Protect internal attributes from new and config calls
    * Dynamically determine width of digital date and time

0.35	- 20 Aug 2013, H.Merijn Brand
    * Indented SYNOPSIS (RT#85697 - Djibril Ousmanou)
    * Support for locales in long day and month formats
    * Support for local offset (in seconds)

0.34	- 28 May 2013, H.Merijn Brand
    * Force update if time wound back (ntp, date command)
    * Upped Copyright to 2013
    * Forced order of config attributes (destroying attribs last)

0.33	- 17 Apr 2012, H.Merijn Brand
    * Upped Copyright to 2012
    * Better protect against CPANTESTERS with non-Tk environment

0.32	- 07 Sep 2011, H.Merijn Brand
    * NAME / DISTNAME in Makefile.PL

0.31	- 07 Sep 2011, H.Merijn Brand
    * More cross-checks for META data

0.30	- 12 May 2011, H.Merijn Brand
    * Initialization fix (attributes can be passed to constructor as it has
      always be intended)
    * Upped Copyright to 2011

0.29	- 02 Aug 2010, H.Merijn Brand
    * Spell-checking
    * Added useInfo, infoColor, infoFormat, and infoFont
    * timeFormat, dateFormat and infoFormat now all support all abbreviations
    * Added timerValue and corresponding Hc, Mc, and Sc format entries

0.28	- 16 Mar 2010, H.Merijn Brand
    * Allow to change backdrop while running
    * Upped Copyright to 2010
    * Dropped YAML spec to 1.0

0.27	- 18 May 2009, H.Merijn Brand
    * Enabled backdrop
    * Add useSecHand option
    * Overhauled the documentation

0.26	- 20 Mar 2009, H.Merijn Brand
    * replace ->_packMethod () with ->manager
    * t/30_dual.t now allows resizing in grid
    * examples/cdclock.pl now acts as I had in mind
    * Added Test::NoWarnings

0.25	- 08 Mar 2009, H.Merijn Brand
    * Vastly improved on resizing.
    * Prefer autoScale => 1 over anaScale => 0
    * Add examples/cdclock.pl (Lightning Talk countdown clock)

0.24	- 20 Jan 2009, H.Merijn Brand
    * Wrong e-mail in META.yml
    * Copyright 2009
    * Guard cpantesters with $DISPLAY unset

0.23	- 21 Oct 2008, H.Merijn Brand
    * Added tgzdist make target
    * Slightly changed the cover make target
    * Allow yyyy in date formats
    * Add `weird' TimeZones to examples/world.pl
    * examples/station.pl can have reversed colors, is resizeable,
      and the ticks don't touch the edges

0.22	- 02 Jan 2008, H.Merijn Brand
    * Tk::Clock is now under git
    * Upped copyright to 2008
    * Added complete prereq list to Makefile.PL

0.21	- 09 Nov 2007, H.Merijn Brand
    * Added tickDiff attribute for width-diff of the ticks
    * Added handCenter attribute (see examples/station.pl)
    * Added examples/station.pl (Stations klok)
    * Added countDown attribute and examples/countdown.pl
    * Fixed auto-resize for clocks with digital disabled
    * Tested with Tk-804.027_501+

0.20	- 08 Oct 2007, H.Merijn Brand
    * Added a 7-clocks wide example/world.pl
    * And fixed the Time Zones in world.pl
    * Raised requirement to 5.006, as I use our and recent Tk builds
      require 5.007

0.19	- 08 May 2007, H.Merijn Brand
    * LICENSE entry for MakeMaker only for newer versions
    * Updated plans/TODO
    * Updated Copyrights
    * Moved pod to bottom
    * Added timeZone support!
    * Split up tests
    * Loosened up the dateFormat and timeFormat attributes to
      allow fixed text. Needs documentation!
    * Added a test with three clocks side-by-side with different TZ's

0.18	- 27 Apr 2007, H.Merijn Brand
    * Added cover make target
    * Added auto-resize test. Still depends on user doing the resize
    * Increased coverage. More coverage is doable, but it'll be a
      booooring job to wait for the tests to finish

0.17	- 16 Nov 2006, H.Merijn Brand
    * Missed one '.' in the format change (K.Wittrock)
    * secsColor was missed in config attrib handling
    * allow both -secsColor and secsColor

0.16	- 17 Oct 2006, H.Merijn Brand
    * Default format for time HH.MM:SS => HH:MM:SS (wish K.Wittrock)
    * Added digiAlign option. Default: "center" (wish K.Wittrock)
    * $clock->coords (...) now used for the hands instead of
      deleting and (re)creating them (suggestion K.Wittrock)

0.15	- 18 Sep 2005, H.Merijn Brand
    * Added Test::Pod
    * Added Test::Pod::Coverage

0.12	- 31 Aug 2005, H.Merijn Brand
    * 24 Hour clock for Abe
    * config () now returns the widget, so it's stackable

0.11	- 14 Aug 2005, H.Merijn Brand
    * Resize reconfigured wrong widget

0.10	- 11 Aug 2005, H.Merijn Brand
    * Solved digital clock misplacement when anaScale passed (Abeltje)
    * Improved resizing
    * Set geometry of parent if MainWindow
    * Changed README and updated Copyright
    * Added a warning for resizing and pack
    * Changed the tests to use Test::More
    * config now returns the widgets new geometry

0.07	- 03 Apr 2000, H.Merijn Brand
    * Added 'w' and 'ww' for weeknumbers in dateFormat
    * Added 'd' and 'dd' for day of the week in timeFormat (for Henry)

0.06	- 27 Sep 1999, H.Merijn Brand
    * Analog clock now scalable (anaScale)
    * Number of ticks configurable (tickFreq)

0.05	- 07 Apr 1999, H.Merijn Brand
    * Updated the README
    * Implemented ddd, dddd, mmm, mmmm (English only)
    * Enabled \n in date format. If combined with empty time
      format, formats like "dddd\nd mmm yyyy" can be used.

0.04	- 15 Dec 1998, H.Merijn Brand
    * Changed README according to perlmodlib manpage
    * Made date formats more Y2K reliable
      dd-mm-yyy still doesn't fit in the allocated space

0.03	- 26 Aug 1998, H.Merijn Brand
    * Analog/Digital part can be disabled.

0.02	- 22 Jul 1998, H.Merijn Brand
    * Support date/time formats through config


Clock.pm
#!/pro/bin/perl

package Tk::Clock;

use strict;
use warnings;

our $VERSION = "0.21";

use Carp;

use Tk::Widget;
use Tk::Derived;
use Tk::Canvas;

use vars qw( @ISA );
@ISA = qw/Tk::Derived Tk::Canvas/;

Construct Tk::Widget "Clock";

my $ana_base = 73;	# Size base for 100%

my %def_config = (
    handColor	=> "Green4",
    secsColor	=> "Green2",
    tickColor	=> "Yellow4",
    tickDiff	=> 0,
    handCenter	=> 0,

    timeZone	=> "",

    timeFont	=> "fixed",
    timeColor	=> "Red4",
    timeFormat	=> "HH:MM:SS",

    dateFont	=> "fixed",
    dateColor	=> "Blue4",
    dateFormat	=> "dd-mm-yy",

    useDigital	=> 1,
    useAnalog	=> 1,
    anaScale	=> 100,
    tickFreq	=> 1,
    ana24hour	=> 0,
    countDown	=> 0,

    digiAlign	=> "center",

    fmtd	=> sub {
	my ($d, $m, $y, $w) = @_;
	sprintf "%02d-%02d-%02d", $d, $m + 1, $y % 100;
	},
    fmtt	=> sub {
	my ($h, $m, $s) = @_;
	sprintf "%02d:%02d:%02d", $h, $m, $s;
	},

    _anaSize	=> $ana_base,	# Default size (height & width)
    _digSize	=> 26,		# Height
    );

sub month ($$)
{
    [[  "1", "01", "Jan", "January"	],
     [  "2", "02", "Feb", "February"	],
     [  "3", "03", "Mar", "March"	],
     [  "4", "04", "Apr", "April"	],
     [  "5", "05", "May", "May"		],
     [  "6", "06", "Jun", "June"	],
     [  "7", "07", "Jul", "July"	],
     [  "8", "08", "Aug", "August"	],
     [  "9", "09", "Sep", "September"	],
     [ "10", "10", "Oct", "October"	],
     [ "11", "11", "Nov", "November"	],
     [ "12", "12", "Dec", "December"	]]->[$_[0]][$_[1]];
    } # month

sub wday ($$)
{
    [[ "Sun", "Sunday"		],
     [ "Mon", "Monday"		],
     [ "Tue", "Tuesday"		],
     [ "Wed", "Wednesday"	],
     [ "Thu", "Thursday"	],
     [ "Fri", "Friday"		],
     [ "Sat", "Saturday"	]]->[$_[0]][$_[1]];
    } # wday

sub min ($$)
{
    $_[0] <= $_[1] ? $_[0] : $_[1];
    } # min

sub max ($$)
{
    $_[0] >= $_[1] ? $_[0] : $_[1];
    } # max

sub resize ($)
{
    my $clock = shift;

    use integer;
    my $data = $clock->privateData;
    my $hght = $data->{useAnalog}  * $data->{_anaSize} +
	       $data->{useDigital} * $data->{_digSize} + 1;
    my $wdth = max ($data->{useAnalog}  * $data->{_anaSize},
		    $data->{useDigital} * 72);
    my $dim  = "${wdth}x${hght}";
    $clock->cget (-height) == $hght &&
     $clock->cget (-width) == $wdth and return $dim;
    if ($clock->parent->isa ("MainWindow")) {
	my $geo = $clock->parent->geometry;
	my @geo = split m/\D/ => $geo;
	if ($geo[0] > 5 && $geo[1] > 5) {
	    $geo =~ s/^\d+x\d+//;
	    $clock->parent->geometry ("$dim$geo");
	    }
	}
    $clock->configure (
	-height => $hght,
	-width  => $wdth);
    $dim;
    } # resize

# Callback when auto-resize is called
sub _resize_auto ($)
{
    my $clock = shift;
    my $data  = $clock->privateData;

    $data->{useAnalog} && $data->{anaScale} == 0 or return;

    my $geo   = $clock->geometry;
    my $owdth = $data->{useAnalog} * $data->{_anaSize};
    my ($gw, $gh) = split m/\D/, $geo;
    $data->{useDigital}   and $gh -= $data->{_digSize};
    my $nwdth = min ($gw, $gh - 1);
    abs ($nwdth - $owdth) > 5 && $nwdth >= 10 or return;

    $data->{_anaSize} = $nwdth - 2;
    $clock->destroyAnalog;
    $clock->createAnalog;
    if ($data->{useDigital}) {
	# Otherwise the digital either overlaps the analog
	# or there is a gap
	$clock->destroyDigital;
	$clock->createDigital;
	}
    $clock->resize;
    } # _resize_auto

sub createDigital ($)
{
    my $clock = shift;

    my $data = $clock->privateData;
    my $wdth = max ($data->{useAnalog}  * $data->{_anaSize},
		    $data->{useDigital} * 72);
    my ($pad, $anchor) = (5, "s");
    my ($x, $y) = ($wdth / 2, $data->{useAnalog} * $data->{_anaSize});
    if    ($data->{digiAlign} eq "left") {
	($anchor, $x) = ("sw", $pad);
	}
    elsif ($data->{digiAlign} eq "right") {
	($anchor, $x) = ("se", $wdth - $pad);
	}
    $clock->createText ($x, $y + $data->{_digSize},
	-anchor	=> $anchor,
	-width  => ($wdth - 2 * $pad),
	-font   => $data->{dateFont},
	-fill   => $data->{dateColor},
	-text   => $data->{dateFormat},
	-tags   => "date");
    $clock->createText ($x, $y + 13,
	-anchor	=> $anchor,
	-width  => ($wdth - 2 * $pad),
	-font   => $data->{timeFont},
	-fill   => $data->{timeColor},
	-text   => $data->{timeFormat},
	-tags   => "time");
#   $data->{Clock_h} = -1;
#   $data->{Clock_m} = -1;
#   $data->{Clock_s} = -1;
    $clock->resize;
    } # createDigital

sub destroyDigital ($)
{
    my $clock = shift;

    $clock->delete ("date");
    $clock->delete ("time");
    } # destroyDigital

sub where ($$$$)
{
    my ($clock, $tick, $len, $anaSize) = @_;      # ticks 0 .. 59
    my ($x, $y, $a);

    $clock->privateData->{countDown} and $tick = (60 - $tick) % 60;
    my $h = ($anaSize + 1) / 2;
    $a = $tick * .104720;
    $x = $len  * sin ($a) * $anaSize / 73;
    $y = $len  * cos ($a) * $anaSize / 73;
    ($h - $x / 4, $h + $y / 4, $h + $x, $h - $y);
    } # where

sub createAnalog ($)
{
    my $clock = shift;

    my $data = $clock->privateData;

    my $h = ($data->{_anaSize} + 1) / 2 - 1;
    my $f = $data->{tickFreq} * 2;
    foreach my $dtick (0 .. 119) {
	$dtick % $f and next;
	my $l = $dtick % 30 == 0 ? $h / 5 :
		$dtick % 10 == 0 ? $h / 8 :
				   $h / 16;
	my $a = ($dtick / 2) * .104720;
	my $x = sin $a;
	my $y = cos $a;
	$clock->createLine (
	    ($h - $l) * $x + $h + 1, ($h - $l) * $y + $h + 1,
	     $h       * $x + $h + 1,  $h       * $y + $h + 1,
	    -tags  => "tick",
	    -arrow => "none",
	    -fill  => $data->{tickColor},
	    -width => $data->{tickDiff} && $dtick % 10 == 0 ? 4.0 : 1.0,
	    );
	}
    $data->{Clock_h} = -1;
    $data->{Clock_m} = -1;
    $data->{Clock_s} = -1;

    $clock->createLine (
	$clock->where (0, 22, $data->{_anaSize}),
	    -tags  => "hour",
	    -arrow => "none",
	    -fill  => $data->{handColor},
	    -width => $data->{_anaSize} / ($data->{handCenter} ? 35 : 26),
	    );
    if ($data->{handCenter}) {
	my $cntr = $data->{_anaSize} /  2;
	my $diam = $data->{_anaSize} / 30;
	$clock->createOval (($cntr - $diam) x 2, ($cntr + $diam) x 2,
	    -tags  => "hour",
	    -fill  => $data->{handColor},
	    -width => 0.
	    );
	}
    $clock->createLine (
	$clock->where (0, 30, $data->{_anaSize}),
	    -tags  => "min",
	    -arrow => "none",
	    -fill  => $data->{handColor},
	    -width => $data->{_anaSize} / ($data->{handCenter} ? 60 : 30),
	    );
    $clock->createLine (
	$clock->where (0, 34, $data->{_anaSize}),
	    -tags  => "sec",
	    -arrow => "none",
	    -fill  => $data->{secsColor},
	    -width => 0.8);
    if ($data->{handCenter}) {
	my $cntr = $data->{_anaSize} /  2;
	my $diam = $data->{_anaSize} / 35;
	$clock->createOval (($cntr - $diam) x 2, ($cntr + $diam) x 2,
	    -tags  => "sec",
	    -fill  => $data->{secsColor},
	    -width => 0.
	    );
	}

    $clock->resize;
    } # createAnalog

sub destroyAnalog ($)
{
    my $clock = shift;

    $clock->delete ("tick");
    $clock->delete ("hour");
    $clock->delete ("min");
    $clock->delete ("sec");
    } # destroyAnalog

sub Populate
{
    my ($clock, $args) = @_;

    my $data = $clock->privateData;
    %$data = %def_config;
    $data->{Clock_h} = -1;
    $data->{Clock_m} = -1;
    $data->{Clock_s} = -1;
    $data->{_time_}  = -1;

    if (ref $args eq "HASH") {
	my %args = %$args;
	foreach my $arg (keys %args) {
	    (my $attr = $arg) =~ s/^-//;
	    exists $data->{$attr} and $data->{$attr} = delete $args{$arg};
	    }
	$args = { %args };
	}

    $clock->SUPER::Populate ($args);

    $clock->ConfigSpecs (
        -width              => [ qw(SELF width              Width              72    ) ],
        -height             => [ qw(SELF height             Height             100   ) ],
        -relief             => [ qw(SELF relief             Relief             raised) ],
        -borderwidth        => [ qw(SELF borderWidth        BorderWidth        1     ) ],
        -highlightthickness => [ qw(SELF highlightThickness HighlightThickness 0     ) ],
        -takefocus          => [ qw(SELF takefocus          Takefocus          0     ) ],
        );

    $data->{useAnalog}  and $clock->createAnalog;
    $data->{useDigital} and $clock->createDigital;
    $clock->resize;

    $clock->repeat (995, ["run" => $clock]);
    } # Populate

sub config ($@)
{
    my $clock = shift;

    ref $clock or croak "Bad method call";
    @_ or return;

    my $conf;
    if (ref $_[0] eq "HASH") {
	$conf = shift;
	}
    elsif (scalar @_ % 2 == 0) {
	my %conf = @_;
	$conf = \%conf;
	}
    else {
	croak "Bad hash";
	}

    my $data = $clock->privateData;
    foreach my $conf_spec (keys %$conf) {
	(my $attr = $conf_spec) =~ s/^-//;
	defined $def_config{$attr} && defined $data->{$attr} or next;
	my $old = $data->{$attr};
	$data->{$attr} = $conf->{$conf_spec};
	if    ($attr eq "tickColor") {
	    $clock->itemconfigure ("tick", -fill => $data->{tickColor});
	    }
	elsif ($attr eq "handColor") {
	    $clock->itemconfigure ("hour", -fill => $data->{handColor});
	    $clock->itemconfigure ("min",  -fill => $data->{handColor});
	    }
	elsif ($attr eq "secsColor") {
	    $clock->itemconfigure ("sec",  -fill => $data->{secsColor});
	    }
	elsif ($attr eq "dateColor") {
	    $clock->itemconfigure ("date", -fill => $data->{dateColor});
	    }
	elsif ($attr eq "dateFont") {
	    $clock->itemconfigure ("date", -font => $data->{dateFont});
	    }
	elsif ($attr eq "timeColor") {
	    $clock->itemconfigure ("time", -fill => $data->{timeColor});
	    }
	elsif ($attr eq "timeFont") {
	    $clock->itemconfigure ("time", -font => $data->{timeFont});
	    }
	elsif ($attr eq "dateFormat") {
	    my %fmt = (
		"d"	=> '%d',	# 6
		"dd"	=> '%02d',	# 06
		"ddd"	=> '%3s',	# Mon
		"dddd"	=> '%s',	# Monday
		"m"	=> '%d',	# 7
		"mm"	=> '%02d',	# 07
		"mmm"	=> '%3s',	# Jul
		"mmmm"	=> '%s',	# July
		"y"	=> '%d',	# 98
		"yy"	=> '%02d',	# 98
		"yyy"	=> '%04d',	# 1998
		"w"	=> '%d',	# 28 (week)
		"ww"	=> '%02d',	# 28
		);
	    my $fmt = $data->{dateFormat};
	    $fmt =~ m{[\%\@\$]} and croak "%, \@ and \$ not allowed in dateFormat";
	    my $xfmt = join "|", reverse sort keys %fmt;
	    my @fmt = split m/\b($xfmt)\b/, $fmt;
	    my $args = "";
	    $fmt = "";
	    foreach my $f (@fmt) {
		if (defined $fmt{$f}) {
		    $fmt .= $fmt{$f};
		    if ($f =~ m/^m+$/) {
			my $l = length ($f) - 1;
			$args .= ", Tk::Clock::month (\$m, $l)";
			}
		    elsif ($f =~ m/^ddd+$/) {
			my $l = length ($f) - 3;
			$args .= ", Tk::Clock::wday (\$wd, $l)";
			}
		    else {
			$args .= ', $' . substr ($f, 0, 1);
			$f =~ m/^y+$/ and
			    $args .= length ($f) < 3 ? " % 100" : " + 1900";
			}
		    }
		else {
		    $fmt .= $f;
		    }
		}
	    $data->{Clock_h} = -1;	# force update;
	    $data->{fmtd} = eval "
		sub
		{
		    my (\$d, \$m, \$y, \$wd, \$w) = \@_;
		    \$w = \$w / 7 + 1;
		    sprintf qq!$fmt!$args;
		    }";
	    }
	elsif ($attr eq "timeFormat") {
	    my %fmt = (
		"H"	=> '%d',	# 6
		"HH"	=> '%02d',	# 06
		"h"	=> '%d',	# 6	AM/PM
		"hh"	=> '%02d',	# 06	AM/PM
		"M"	=> '%d',	# 7
		"MM"	=> '%02d',	# 07
		"S"	=> '%d',	# 45
		"SS"	=> '%02d',	# 45
		"A"	=> '%s',	# PM
		"dd"	=> '%.2s',	# Mo
		"ddd"	=> '%.3s',	# Mon
		"dddd"	=> '%s',	# Monday
		);
	    my $fmt = $data->{timeFormat};
	    $fmt =~ m/[\%\@\$]/ and croak "%, \@ and \$ not allowed in timeFormat";
	    my $xfmt = join "|", reverse sort keys %fmt;
	    my @fmt = split m/\b($xfmt)\b/, $fmt;
	    my $arg = "";
	    $fmt = "";
	    foreach my $f (@fmt) {
		if ($f =~ m/^dd{1,3}$/) {
		    $fmt .= $fmt{$f};
		    $arg .= ", Tk::Clock::wday (\$d, 1)";
		    }
		elsif (defined $fmt{$f}) {
		    $fmt .= $fmt{$f};
		    $arg .= ', $' . substr ($f, 0, 1);
		    }
		else {
		    $fmt .= $f;
		    }
		}
	    $data->{fmtt} = eval "
		sub
		{
		    my (\$H, \$M, \$S, \$d) = \@_;
		    my \$h = \$H % 12;
		    my \$A = \$H > 11 ? 'PM' : 'AM';
		    sprintf qq!$fmt!$arg;
		    }";
	    }
	elsif ($attr eq "tickFreq") {
#	    $data->{tickFreq} < 1 ||
#	    $data->{tickFreq} != int $data->{tickFreq} and
#		$data->{tickFreq} = $old;
	    unless ($data->{tickFreq} == $old) {
		$clock->destroyAnalog;
		$clock->createAnalog;
		}
	    }
	elsif ($attr eq "anaScale") {
	    $data->{anaScale} eq "auto" and $data->{anaScale} = 0;
	    if ($data->{anaScale} == 0) {	# 0 will be auto some time
		$clock->Tk::bind         ("Tk::Clock","<<ResizeRequest>>", \&_resize_auto);
		$clock->parent->Tk::bind ("<<ResizeRequest>>", \&_resize_auto);
		$clock->_resize_auto;
		}
	    else {
		my $new_size = int ($ana_base * $data->{anaScale} / 100.);
		unless ($new_size == $data->{_anaSize}) {
		    $data->{_anaSize} = $new_size;
		    $clock->destroyAnalog;
		    $clock->createAnalog;
		    if (exists $conf->{anaScale} && $data->{useDigital}) {
			# Otherwise the digital either overlaps the analog
			# or there is a gap
			$clock->destroyDigital;
			$clock->createDigital;
			}
		    $clock->after (5, ["run" => $clock]);
		    }
		}
	    }
	elsif ($attr eq "useAnalog") {
	    if    ($old == 1 && $data->{useAnalog} == 0) {
		$clock->destroyAnalog;
		$clock->destroyDigital;
		$data->{useDigital} and $clock->createDigital;
		}
	    elsif ($old == 0 && $data->{useAnalog} == 1) {
		$clock->destroyDigital;
		$clock->createAnalog;
		$data->{useDigital} and $clock->createDigital;
		}
	    $clock->after (5, ["run" => $clock]);
	    }
	elsif ($attr eq "useDigital") {
	    if    ($old == 1 && $data->{useDigital} == 0) {
		$clock->destroyDigital;
		}
	    elsif ($old == 0 && $data->{useDigital} == 1) {
		$clock->createDigital;
		}
	    $clock->after (5, ["run" => $clock]);
	    }
	elsif ($attr eq "digiAlign") {
	    if ($data->{useDigital} && $old ne $data->{digiAlign}) {
		$clock->destroyDigital;
		$clock->createDigital;
		$clock->after (5, ["run" => $clock]);
		}
	    }
	}
    $clock->resize;
    $clock;
    } # config

sub run ($)
{
    my $clock = shift;

    my $data = $clock->privateData;

    $data->{timeZone} and local $ENV{TZ} = $data->{timeZone};
    my $t = time;
    $t == $data->{_time_} and return;	# Same time, no update
    $data->{_time_} = $t;
    my @t = localtime $t;

    unless ($t[2] == $data->{Clock_h}) {
	$data->{Clock_h} = $t[2];
	$data->{useDigital} and
	    $clock->itemconfigure ("date",
		-text => &{$data->{fmtd}} (@t[3,4,5,6,7]));
	}

    unless ($t[1] == $data->{Clock_m}) {
        $data->{Clock_m} = $t[1];
	if ($data->{useAnalog}) {
	    my ($h24, $m24) = $data->{ana24hour} ? (24, 2.5)  : (12, 5);
	    $clock->coords ("hour",
		$clock->where (($data->{Clock_h} % $h24) * $m24 + $t[1] / $h24, 22, $data->{_anaSize}));

	    $clock->coords ("min",
		$clock->where ($data->{Clock_m}, 30, $data->{_anaSize}));
	    }
	}

    $data->{Clock_s} = $t[0];
    if ($data->{useAnalog}) {
	$clock->coords ("sec",
	    $clock->where ($data->{Clock_s}, 34, $data->{_anaSize}));
	}
    $data->{useDigital} and
	$clock->itemconfigure ("time",
	    -text => &{$data->{fmtt}} (@t[2,1,0,6]));

    $data->{anaScale} == 0 and $clock->_resize_auto;
    } # run

1;

__END__

=head1 NAME

Tk::Clock - Clock widget with analog and digital display

=head1 SYNOPSIS

use Tk
use Tk::Clock;

$clock = $parent->Clock (?-option => <value> ...?);

$clock->config (	# These reflect the defaults
    useDigital	=> 1,
    useAnalog	=> 1,
    anaScale	=> 100,
    ana24hour	=> 0,
    handColor	=> "Green4",
    secsColor	=> "Green2",
    handCenter	=> 0,
    tickColor	=> "Yellow4",
    tickFreq	=> 1,
    tickDiff    => 0,
    timeZone	=> "",
    timeFont	=> "fixed",
    timeColor	=> "Red4",
    timeFormat	=> "HH:MM:SS",
    dateFont	=> "fixed",
    dateColor	=> "Blue4",
    dateFormat	=> "dd-mm-yy",
    digiAlign   => "center",
    );

=head1 DESCRIPTION

Create a clock canvas with both an analog- and a digital display. Either
can be disabled by setting useAnalog or useDigital to 0 resp.

Legal dateFormat characters are d and dd for date, ddd and dddd for weekday,
m, mm, mmm and mmmm for month, y and yy for year, w and ww for weeknumber and
any separators :, -, / or space.

Legal timeFormat characters are H and HH for hour, h and hh for AM/PM hour,
M and MM for minutes, S and SS for seconds, A for AM/PM indicator, d and dd
for day-of-the week in two or three characters resp. and any separators :,
-, . or space.

Meaningful values for tickFreq are 1, 5 and 15 showing all ticks, tick
every 5 minutes or the four main ticks only, though any positive integer
will do (put a tick on any tickFreq minute). When setting tickDiff to a
true value, the major ticks will use a thicker line than the minor ticks.

The analog clock can be enlaged or reduced using anaScale for which the
default of 100% is about 72x72 pixels. Setting anaScale to 0, will try to
resize the widget to it's container automatically.

For digiAlign, "left", "center", and "right" are the only supported values.
Any other value will be interpreted as the default "center".

When using C<pack> for your geometry management, be sure to pass
C<-expand =&gt; 1, -fill =&gt; "both"> if you plan to resize with
C<anaScale> or enable/disable either analog or digital after the
clock was displayed.

=head1 BUGS

If the system load's too high, the clock might skip some seconds.

Due to the fact that the year is expressed in 2 digit's, this
widget is not Y2K compliant in the default configuration.

There's no check if either format will fit in the given space.

=head1 TODO

* Using POSIX' strftime () for dateFormat. Current implementation
  would probably make this very slow.
* Full support for multi-line date- and time-formats with auto-resize.
* Countdown clock API, incl action when done.
* Better docs for the attributes

=head1 AUTHOR

H.Merijn Brand <h.m.brand@xs4all.nl>

Thanks to Larry Wall for inventing perl.
Thanks to Nick Ing-Simmons for providing perlTk.
Thanks to Achim Bohnet for introducing me to OO (and converting
    the basics of my clock.pl to Tk::Clock.pm).
Thanks to Sriram Srinivasan for understanding OO though his Panther book.
Thanks to all CPAN providers for support of different modules to learn from.
Thanks to all who have given me feedback.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 1999-2007 H.Merijn Brand

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
